### ISSUE TITLE
Dynaform support of tables defined in app_startup

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1664]
_Imported from trac ticket 1664.
Created by **kbower**
Opened in trac: May 23, 2012, 12:23:35 PM
Last modified in trac: Aug 11, 2012, 3:46:07 PM_
***
Funny you ask... I nearly began this a couple weeks back (the ability to have a dynaform that gets its transferobject metadata from a service request instead of metadata).  It's a bit involved, but something I think we'll need.  I thought I was going to need it a few weeks back.  I was also going to restructure the DynaForm table because we can practically automate it if we simplify this:

```python
data.append({
    'class': DynaForm,
    'columns': ('id','label','permissionname','type','transferobject','event',),
    'values': (
        ('sites','Sites','adminsites','dynaform','com.rarch.transferObjects.Site',None,),
        ('sitegrp','SiteGroups','adminsitegroup','dynaform','com.rarch.transferObjects.SiteGroup',None,),
        ('sitephone','SitePhones','adminsitephone','dynaform','com.rarch.transferObjects.SitePhone',None,),
        ('fillseq','FillSequences','adminfillseq','dynaform','com.rarch.transferObjects.FillSequence',None,),
        ('loc','Locations','adminloc','dynaform','com.rarch.transferObjects.Location',None,),
        ('zone','Zones','adminzone','dynaform','com.rarch.transferObjects.Zone',None,),
...
        ('installments','InstallmentContracts','installmentdetail','event',None,'InstallmentQueryNav',),
        ('creditapplication','CreditApplication','creditapplicationperm','event',None,'CreditApplicationQuery',),
```
Into this:

```python
data.append({
    'class': DynaForm,
    'columns': ('class','event',),
    'values': (
        ('Site',None,),
        ('SiteGroup',None,),
        ('SitePhone',None,),
        ('FillSequence',None,),
        ('Location',None,),
        ('Zone',None,),
...
        ('InstallmentContract','InstallmentQueryNav',),
        ('CreditApplication','CreditApplicationQuery',),
```

All the rest of the data in that table is superfluous.
'class' becomes the 'id'
'label' is the class with an 's' at the end
'permissionname' should be structured for consistency as 'admin' + lowercase(class)
'type' is implied by whether 'event' is null or not
'transferobject' is constructed from 'class'... it is doesn't exist then a service is invoked to get it dynamically (and cache the metadata, of course, so it's only needed once)

This would make datasetup greatly simplified.

On 5/23/2012 10:23 AM, Rob Neville wrote:

[[Image(None)]]

***
* kbower attached [None](https://trac.retailarchitects.com/trac/attachment/ticket/1664/None) on May 23, 2012, 12:23:35 PM

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 11, 2012, 3:46:07 PM:_

(Changeset [[2971]](../commit/c288c26756ac36a3ff9362ae35c9abe2bf0b691a))
* Add new system rule called 'model_customizations' for end user database modifications. (closes #1664)
  * Customers can add columns or tables in this rule, which will automatically create them if they are needed.
  * Flex client requests this additional metadata from the server at startup and appends it to the metadata it introspects from flex classes
  * controllers for custom classes are dynamically generated
* Remove AttributeTypes.as and AttributeType.as classes and merge their usage with TransObjMetadata.as (both were essentially metadata introspection mechanisms)
* update TransferObjectParser.castResultAs() to accept either a concrete flex Class or a TransObjMetadata object describing the class to cast
* Make it easier to specify that a relationship's code should be auto generated for flex: just add "flexcode=True" to the relationship in mappings.py 
* add "info" dictionary for relationship() and column_property(), similar to Column and Table in purpose
* Refactor (simplify) Dynaform data setup.
  * Renamed DynaForm to MenuItem (they are not all dynaforms, the event driven menu items do something else)
  * Renamed DynaMenu to Menu
  * Hierarchy now: Menu=>MenuCategory=>MenuItem
  * renamed "id" columns of these tables to menuid, menucategoryid, menuitemid  
* To add new DynaForm (also may do so in rules for custom tables):
  * Add row to MenuItem, e.g., ('PromotionPrice', None),
  * Link in MenuCategoryMenuItem, e.g.,  ('promotions','PromotionPrice',),
  * Add permission for MenuItem in format: 'menu' + lowercase(menuitemid, which is classname), e.g., ('menupromotionprice','PRODUCTS - Ability to create/edit promotion prices.'),
* Add ability to completely disable certain menu categories or items so they don't show up ('enabled' column)
* Remove DynaFormAssociation and DynaFormCollection (and tables from database).  These are not user defined, but metadata driven and sqlalchemy knows better than user.
  * To specify a relationship should be included in the dynaform as a related collection:
    * Just mark the relationship in mappings.py (or model_customizations rule) as "dynaform_relationship=True"
    * It must be collection relationship instead of scalar (for now, at least)
      * For now, many to many is best (required, unless read only)
    * It probably should be in merge cascade, or will effectively be read only (some potential there)
    * It will need to be fetched at get time (added to get_eagerloads)
    * add label to labels.properties (or info dictionary in case of 'model_customizations' rule)
    * add a permission in the format 'set' + lowercase(class name) + relationship name, e.g. 'setgrouppermissions'
* Fix to enforce permission governing associating relationships from dynaforms (fixes [#1861](1861))
* Standardize the naming of the admin permission needed to edit/create records of a dynaform class:
  * By default, use the same permission as the menu permission needed for the dynaform class, which is "menu" + lowercase(classname)
  * If more granularity is required, add 'explicitpermissions':True to the table's info dictionary, in which case three permissions are used:
    * "query" + lowercase(classname)
    * "delete" + lowercase(classname)
    * "edit" + lowercase(classname)  future: what about "create"?
* Fix dynaform associated relations datagrid to only add column once and also to always add primarykey columns
* rename token "flexserialize"=>"clientserialize" for specifying whether a client should serialize a column in tables.py
* Make passed rule parameters override pylotengine.model.objects globals() variables of same name, like a computer language should (needed for customization_framework)
* database upgrade fixes:
  * properly deal with situation where constraint was already dropped and is then needed to be dropped again for column or table
  * dependencies: add primary/unique keys _before_ adding new tables in case they have fks to them
  * when dropping/creating constraints, we temporarily add a migrate.changeset.constraint to the table which knows how to drop or add itself. The fix included is to remove that after adding/dropping so it doesn't mess up something later (e.g. a table drop)
* add copypermission() function for dataupdate.py use  
* use `[Fieldinfo(advancedQuery="false")]` instead of `[Dynamicinfo(skiprelations="true")]` 
* move Kane's custom metadata to 'model_customizations' rule and add dynaforms for them

**Comment by kbower** [comment]
_**kbower** trac status change on Aug 11, 2012, 3:46:07 PM:_ **closed**
